### PR TITLE
feat(ingest/s3): Partition support

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

## Summary by CodeRabbit

- **New Features**
  - Enhanced documentation for S3 dataset ingestion, including flexible path configuration and partition auto-detection.
  - Added new attributes for data ingestion configuration, improving control over partition handling and metadata management.
  - Introduced a new `Folder` class to better manage folder metadata for partitioned datasets.
  - New JSON configurations for S3 data source integration with advanced partition detection options.

- **Bug Fixes**
  - Restructured JSON schemas for improved clarity and consistency in metadata representation across various test cases.

- **Tests**
  - Improved validation logic in S3 integration tests to ensure uploaded files match expected values.

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 1]
```diff
@@ -753,23 +926,86 @@ def get_dir_to_process(
...
+        partition: Optional[str] = None,
+    ) -> List[Folder]:
+        prefix_to_list = prefix
+        if partition:
+            prefix_to_list = f"{prefix}/{partition}"
+        files = list(
+            bucket.objects.filter(Prefix=f"{prefix_to_list}").page_size(PAGE_SIZE)
+        )
+        files = sorted(files, key=lambda a: a.last_modified)
+        grouped_files = groupby(files, lambda x: x.key.rsplit("/", 1)[0])
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/s3/source.py:964]: Suggestion is based on usage of `os.sep` elsewhere in the code. When do we need to use `os.sep` vs `"\"` ? Please look for any other places where we need to use `os.sep` vs `"\"`.

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 2]
```diff
@@ -753,23 +926,86 @@ def get_dir_to_process(
...
+        return [f"{protocol}{bucket_name}/{folder}"]
+
+    def get_partitions(
+        self,
+        path_spec: PathSpec,
+        bucket: Any,  # Todo: proper type
+        prefix: str,
+        partition: Optional[str] = None,
+    ) -> List[Folder]:
+        prefix_to_list = prefix
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:957]: Please add description and examples to this method.

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 3]
```diff
@@ -753,23 +926,86 @@ def get_dir_to_process(
...
+            else:
+                return [f"{protocol}{bucket_name}/{folder}"]
+        return [f"{protocol}{bucket_name}/{folder}"]
+
+    def get_partitions(
+        self,
+        path_spec: PathSpec,
+        bucket: Any,  # Todo: proper type
+        prefix: str,
+        partition: Optional[str] = None,
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:955]: Looks like `partition` argument is not passed in codebase. Please remove if not required, in favour of simplicity.

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 4]
```diff
@@ -753,23 +926,86 @@ def get_dir_to_process(
...
+                    size=file_size,
+                )
+            )
+
+        return partitions
 
     def s3_browser(
         self, path_spec: PathSpec, sample_size: int
-    ) -> Iterable[Tuple[str, datetime, int]]:
+    ) -> Iterable[Tuple[str, datetime, int, List[Folder]]]:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:1008]: Ideally, we should create a dataclass for return type.

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py [hunk 5]
```diff
@@ -28,6 +30,57 @@
...
+        description="The date format to use when sorting. This is used to parse the date from the key. The format should follow the java [SimpleDateFormat](https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html) format.",
+    )
+
+    @pydantic.validator("date_format", always=True)
+    def convert_date_format_to_python_format(cls, v: Optional[str]) -> Optional[str]:
+        if v is None:
+            return None
+        else:
+            for java_format, python_format in java_to_python_mapping.items():
+                v = v.replace(java_format, f"%{python_format}")
```
[the remark this answers is not in the document]
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:75]: In the future, this should be supported across platforms (java, python, etc..) and I felt like the java one is more common

File: metadata-ingestion/src/datahub/ingestion/source/s3/config.py [hunk 6]
```diff
@@ -48,6 +48,11 @@ class DataLakeSourceConfig(
         description="Whether or not to create tags in datahub from the s3 object",
     )
 
+    get_all_partitions: bool = Field(
+        default=False,
+        description="Whether to list all partitions in the table, or only the latest",
+    )
```
[the remark this answers is not in the document]
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/s3/config.py:54]: you are right it left there accidentally

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 7]
```diff
@@ -590,6 +617,39 @@ def _create_table_operation_aspect(self, table_data: TableData) -> OperationClas
...
+        if max_partition_id is not None:
+            max_partition_summary = PartitionSummaryClass(
+                partition=max_partition_id,
+                createdTime=int(max_partition.creation_time.timestamp() * 1000),
+                lastModifiedTime=int(
+                    max_partition.modification_time.timestamp() * 1000
+                ),
+            )
+
+            min_partition_summary: Optional[PartitionSummaryClass] = None
```
[the remark this answers is not in the document]
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/s3/source.py:638]: true, thanks

File: metadata-ingestion/tests/unit/s3/test_s3_source.py [hunk 8]
```diff
@@ -70,6 +72,75 @@ def test_path_spec():
...
+        ),
+    ],
+)
+def test_path_spec_partition_detection(
+    path_spec: str, path: str, expected: List[Tuple[str, str]]
+) -> None:
+    ps = PathSpec(include=path_spec, default_extension="csv", allow_double_stars=True)
+    assert ps.allowed(path)
+    partitions = ps.get_partition_from_path(path)
+    assert partitions == expected
```
Comment (mayurinehate) [inline metadata-ingestion/tests/unit/s3/test_s3_source.py:141]: Thank you for adding these.

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py [hunk 9]
```diff
@@ -358,23 +360,45 @@ def extract_variable_names(self):
         return matches
 
     def get_partition_from_path(self, path: str) -> Optional[List[Tuple[str, str]]]:
+        # Automatic partition detection supports four methods to get partiton keys and values from path:
+        # Let's say we have the following path => year=2024/month=10/day=11 for this example you can specify the following path spec expressions:
+        #   1. User can specify partition_key and partition_value in the path like => {partition_key[0]}={partition_value[0]}{partition_key[1]}={partition_value[1]}{partition_key[2]}={partition_value[2]}
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:365]:
```suggestion
        #   1. User can specify partition_key and partition_value in the path like => {partition_key[0]}={partition_value[0]}/{partition_key[1]}={partition_value[1]}/{partition_key[2]}={partition_value[2]}
```

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py [hunk 10]
```diff
@@ -358,23 +360,45 @@ def extract_variable_names(self):
...
+        #   2. User can specify only partition key and the partition key will be used as partition name like => year={year}month={month}day={day}
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:366]:
```suggestion
        #   2. User can specify only partition key and the partition key will be used as partition name like => year={year}/month={month}/day={day}
```

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 11]
```diff
@@ -153,6 +153,14 @@ def partition_id_text(self) -> Optional[str]:
         return None
 
 
+@dataclasses.dataclass
+class BrowsePath:
+    file: str
+    timestamp: datetime
+    size: int
+    partitions: List[Folder]
+
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:162]: Nice! So browser would browse through file system to yield one BrowsePath for one identified table and that table would then be emitted as dataset if its allowed by patterns, right ?

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 12]
```diff
@@ -868,24 +812,41 @@ def get_dir_to_process(
...
+        - Creation time is the earliest creation time of all files in the folder.
+        - Modification time is the latest modification time of all files in the folder.
+        - Size is the sum of all file sizes in the folder.
+        - Sample file path is used for schema inference later. (sample file is the latest created file in the folder)
+
+        Parameters:
+        path_spec (PathSpec): The path specification used to determine partitioning.
+        bucket (Any): The S3 bucket object.
+        prefix (str): The prefix path in the S3 bucket to list objects from.
+        partition (Optional[str]): An optional partition string to append to the prefix.
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:843]: Can remove partition from here now that its removed from method

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py [hunk 13]
```diff
@@ -263,6 +509,44 @@ def _extract_table_name(self, named_vars: dict) -> str:
             raise ValueError("path_spec.table_name is not set")
         return self.table_name.format_map(named_vars)
 
+    def extract_datetime_partition(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:512]: Is this used anywhere ? Would be good to add unit tests for this too.

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 14]
```diff
@@ -753,23 +926,86 @@ def get_dir_to_process(
...
+                )
+                continue
+
+            id = path_spec.get_partition_from_path(
+                self.create_s3_path(max_file.bucket_name, max_file.key)
+            )
+            partitions.append(
+                Folder(
+                    partition_id=id,
+                    is_partition=True if id else False,
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/s3/source.py:886]: Should we set empty partitions `[]` if `id` is absent ?

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 15]
```diff
@@ -737,24 +937,68 @@ def s3_browser(
...
+                                == FolderTraversalMethod.MIN_MAX
+                            ):
+                                dirs_to_process_min = self.get_dir_to_process(
+                                    bucket_name=bucket_name,
+                                    folder=f + "/",
+                                    path_spec=path_spec,
+                                    protocol=protocol,
+                                    min=True,
+                                )
+                                dirs_to_process.append(dirs_to_process_min[0])
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:973]: Should we push down traversal mode as max or min_max to `get_dirs_to_process` ? Without that, `list_folders` would be done twice ?

**Review by mayurinehate — COMMENTED** [review]
Looking good.

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py (hunk 11 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/s3/source.py:162]: yes, exactly

**Review by mayurinehate — APPROVED** [review]
LGTM

**Review by jjoyce0510 — APPROVED** [review]

**Review by jjoyce0510 — APPROVED** [review]

**Pull request merged**